sqlmesh/core/analytics/collector.py

```diff
@@ -0,0 +1,351 @@
+            "RUN_START",
+            {
+                "run_id": run_id,
+                "engine_type": engine_type.lower(),
+                "state_sync_type": state_sync_type.lower(),
+            },
+        )
+
+    def on_run_end(self, *, run_id: str, error: t.Optional[t.Any] = None) -> None:
+        """Called after a run ends.
+
+        Args:
+            run_id: The ID of the run.
+            error: The error that occurred during the run, if any.
+        """
+        self._add_event(
+            "RUN_END",
+            {
+                "run_id": run_id,
+                "succeeded": error is None,
+                "error": type(error).__name__ if error else None,
+            },
+        )
+
+    def on_migration_end(
+        self,
+        *,
+        from_sqlmesh_version: str,
+        state_sync_type: str,
+        migration_time_sec: float,
+        error: t.Optional[t.Any] = None,
+    ) -> None:
+        """Called after the migration of the SQLMesh state ends.
+
+        Args:
+            from_sqlmesh_version: The version of SQLMesh from which the migration started.
+            state_sync_type: The type of the engine used to store the SQLMesh state.
+            migration_time_sec: The time it took to migrate the SQLMesh state in (fractional) seconds.
+            error: The error that occurred during the migration, if any.
+        """
+        self._add_event(
+            "MIGRATION_END",
+            {
+                "from_sqlmesh_version": from_sqlmesh_version,
+                "state_sync_type": state_sync_type,
+                "succeeded": error is None,
+                "error": type(error).__name__ if error else None,
+                "migration_time_ms": int(migration_time_sec * 1000),
+            },
+        )
+
+    def flush(self) -> None:
+        """Flushes the events to the dispatcher."""
+        self._dispatcher.flush()
+
+    def shutdown(self, flush: bool = True) -> None:
+        """Shuts down the collector."""
+        self._dispatcher.shutdown(flush=flush)
+
+    def _on_command(
+        self,
+        event_type: str,
+        command_name: str,
+        command_args: t.Collection[str],
+        **kwargs: t.Any,
+    ) -> None:
+        event = {
+            "command_name": command_name,
+            "command_args": list(command_args),
+            **kwargs,
+        }
+        self._add_event(event_type, event)
+
+    def _add_event(self, event_type: str, event: t.Dict[str, t.Any]) -> None:
+        self._dispatcher.add_event(
+            {
+                "user_id": self._user.id,
+                "process_id": self._process_id,
+                "seq_num": next(self._seq_num),
+                "event_type": event_type,
+                "client_ts": now_timestamp(),
+                "event": json.dumps(event),
+            }
+        )
+
+    @cached_property
+    def _user(self) -> User:
+        return User.load_or_create(self._sqlmesh_path)
+
+
+class User(PydanticModel):
```

Comment:
there's no reason for this to be a pydantic model

Reply:
What do you mean? What do you suggest to do instead?

Reply:
it can just be a data class

Reply:
I also serialize it into a dict before dumping into YAML. Not sure I want to introduce new patterns here.